src/python/pants/backend/cargo_bootstrap/subsystems/rustup.py

```diff
@@ -0,0 +1,290 @@
+    register('--rust-components', type=list_option, default=[
+      'rustfmt-preview',
+      'rust-src',
+      'clippy-preview',
+    ], help='Components of the rust toolchain to install.')
+    register('--cargo-version', type=str, default='0.2.1',
+             help='Version of the cargo build tool to use.')
+
+  @memoized_property
+  def update_request(self):
+    opts = self.get_options()
+    version = opts.toolchain_version
+    components = opts.rust_components
+    return RustUpdateRequest(
+      toolchain_root=self._toolchain_root,
+      toolchain_version=text_type(version),
+      rust_components=tuple(components),
+    )
+
+  @memoized_property
+  def cargo_version(self):
+    return self.get_options().cargo_version
+
+  @memoized_property
+  def _toolchain_root(self):
+    return os.path.join(get_pants_cachedir(), 'rust')
+
+  @memoized_property
+  def _cargo_home(self):
+    return os.path.join(self._toolchain_root, 'cargo')
+
+  @memoized_property
+  def _rustup_home(self):
+    return os.path.join(self._toolchain_root, 'rustup')
+
+  @memoized_property
+  def _cargo_bin_dir(self):
+    return os.path.join(self._cargo_home, 'bin')
+
+  @memoized_property
+  def pants_owned_rustup_path(self):
+    return os.path.join(self._cargo_bin_dir, 'rustup')
+
+  @memoized_property
+  def pants_owned_rustup_globs(self):
+    return PathGlobsAndRoot(PathGlobs(['rustup']), self._cargo_bin_dir)
+
+  @memoized_property
+  def cargo_ensure_installed_path(self):
+    return os.path.join(self._cargo_bin_dir, 'cargo-ensure-installed')
+
+  @memoized_property
+  def cargo_ensure_installed_globs(self):
+    return PathGlobsAndRoot(PathGlobs(['cargo-ensure-installed']), self._cargo_bin_dir)
+
+  @memoized_property
+  def _cmake(self):
+    return CMakeForGRPC.scoped_instance(self)
+
+  @memoized_property
+  def _go_dist(self):
+    return GoForGRPC.scoped_instance(self)
+
+  @memoized_property
+  def _protoc(self):
+    return ProtocForGRPC.scoped_instance(self)
+
+  @memoized_method
+  def rustup_exec_env(self):
+    return {
+      'CARGO_HOME': self._cargo_home,
+      'RUSTUP_HOME': self._rustup_home,
+      'PATH': create_path_env_var([
+        self._cmake.bin_dir(),
+        self._go_dist.bin_dir(),
+        self._protoc.bin_dir(),
+        # TODO: execute this in a sealed path!
+      ], env=os.environ.copy(), prepend=True),
+    }
+
+
+class RustupExe(datatype([('exe', Snapshot)])): pass
+
+
+@rule(RustupExe, [Select(Rustup)])
+def unpack_rustup(rustup):
+  if not is_executable(rustup.pants_owned_rustup_path):
+    script = rustup.hackily_snapshot(context=None)
+    req = ExecuteProcessRequest(
+      argv=('./rustup.sh', '-y', '--no-modify-path', '--default-toolchain', 'none'),
+      input_files=script.directory_digest,
+      description='execute rustup',
+      env=rustup.rustup_exec_env(),
+    )
+    # TODO: this isn't remotable -- it will modify the local filesystem at CARGO_HOME and
```

Comment:
This is a blocker... although nothing is preventing access to `CWD` here, we definitely should not be relying on that access.

So the big pieces here (and I think that they're important to plan/design/hopefully-implement before landing this in order to actually prove out the assumptions of the model):
1. Using `PathGlobsAndRoot` from within an `@rule` is not currently kosher because we cannot invalidate dependencies on files outside of the workdir.
2. The right way to get files _out_ of a `Snapshot` and into a local working copy is via https://github.com/pantsbuild/pants/blob/master/src/python/pants/engine/scheduler.py#L332-L347... but that is not exposed to `@rules`, and shouldn't be. It could _probably_ be exposed to `@console_rules`, with a bit of design work.

Reply:
For 2, we can just do this in the v1 `Task` for now, and worry about the design work later :)

Reply:
Fine with that.